Glossary update: add architect terms and xrefs to new pages (#26963)
angular/angular: aio/content/guide/glossary.md

A structure that provides metadata for a class. See [decorator](guide/glossary#decorator).

{@a architect}

## Architect

> The tool that the CLI uses to perform complex tasks such as compilation and test running, according to provided configurations.  Architect is a shell that runs a specific *builder* (an [npm package](#npm-package)) for a given *target*, using a builder-specific configuration.

Author: Added ##builder

> The tool that the CLI uses to perform complex tasks such as compilation and test running, according to provided configurations.  Architect is a shell that runs a specific *builder* (an [npm package](#npm-package)) for a given *target*, using a builder-specific configuration.

Author: The positioning is not clear to me either. Or to Hans, who has agreed that the terms are a bit slippery.

For example, the builder for the `lint` target is [TSLint](https://www.npmjs.com/package/tslint), and the configuration is specified in a `tslint.json` file. For more information on available targets, see [Project tool configuration](guide/workspace-config#project-configuration-options).

> You can invoke Architect for a given target using the [CLI command `ng run`](cli/run).

Author: Architect invokes a builder. The builder invokes a tool, using a target-specific configuration. The tool creates the targeted result -- that can be anything you define in your builder.

This can improve the user experience by quickly launching a static rendered page (a skeleton common to all pages) while the browser downloads the full client version and switches to it automatically after the code loads.

See also [Service Worker and PWA](guide/service-worker-intro).
{@a architect}

## Architect

The tool that the CLI uses to perform complex tasks such as compilation and test running, according to a provided configuration.
Architect is a shell that runs a [builder](#builder) (defined in an [npm package](#npm-package)) with a given [target configuration](#target).

In the [workspace configuration file](guide/workspace-config#project-tool-configuration-options), an   "architect" section provides configuration options for Architect builders.

For example, a built-in builder for linting is defined in the package `@angular-devkit/build_angular:tslint`, which uses the [TSLint](https://palantir.github.io/tslint/) tool to perform linting, with a configuration specified in a `tslint.json` file.

> Use the [CLI command `ng run`](cli/run) to invoke a builder with a specific configuration.

Reviewer: ```suggestion
Use the [CLI command `ng run`](cli/run) to invoke a builder by specifying a [target configuration](#target) associated with the builder.
```

Learn more in [Bootstrapping](guide/bootstrapping).

{@a builder}

## builder

A function that uses the [Architect](#architect) API to perform a complex process such as "build" or "test".

> The builder code is defined in an [npm package](#npm-package).

Reviewer: ```suggestion
The code for a builder is packaged and can be published in an [npm package](#npm-package).
```

Reviewer: Any questions about this?

Reviewer 2: lgtm

For example, [BrowserBuilder](https://github.com/angular/angular-cli/tree/master/packages/angular_devkit/build_angular/src/browser) runs a [webpack](https://webpack.js.org/) build for a browser target and [KarmaBuilder](https://github.com/angular/angular-cli/tree/master/packages/angular_devkit/build_angular/src/karma) starts the Karma server and runs a webpack build for unit tests.

The [CLI command `ng run`](cli/run) invokes a builder with a specific [target configuration](#target).

> The [workspace configuration](guide/workspace-config) file, `angular.json`, contains default configurations for built-in builders.

Reviewer: ```suggestion
The [workspace configuration](guide/workspace-config) file, `angular.json`, contains default [target configurations](#target) for builders provided by [Architect](#architect).
```

Reviewer: Any questions about this?

Reviewer 2: lgtm

A class with the `@Component()` [decorator](guide/glossary#decorator) that associates it with a companion [template](guide/glossary#template). Together, the component and template define a [view](guide/glossary#view).
A component is a special type of [directive](guide/glossary#directive).
The `@Component()` decorator extends the `@Directive()` decorator with template-oriented features.

An Angular component class is responsible for exposing data and handling most of the view's display and user-interaction logic through [data binding](guide/glossary#data-binding).

Read more about components, templates, and views in [Architecture Overview](guide/architecture).

{@a cli-config}

{@a config}

## configuration

A file named `angular.json` at the root level of an Angular [workspace](#workspace) provides workspace-wide and project-specific configuration defaults for build and development tools provided by the [Angular CLI](#cli).

For more information, see [Workspace Configuration](guide/workspace-config).

> Additional project-specific configuration files are used by subsystems and tools, such as `package.json` for the [npm package manager](#npm-package), and `tsconfig.json` for [TypeScript transpilation](#transpile), and `tslint.json` for [TSLint](https://palantir.github.io/tslint/).

Reviewer 3: I like this :-)

## component

A class with the `@Component()` [decorator](#decorator) that associates it with a companion [template](#template). Together, the component and template define a [view](#view).
A component is a special type of [directive](#directive).
The `@Component()` decorator extends the `@Directive()` decorator with template-oriented features.

An Angular component class is responsible for exposing data and handling most of the view's display and user-interaction logic through [data binding](#data-binding).

Read more about components, templates, and views in [Architecture Overview](guide/architecture).

{@a cli-config}

{@a config}

> ## configuration

Reviewer 2: isn't this too generic? how about "Workspace configuration"?

Read more about components, templates, and views in [Architecture Overview](guide/architecture).

{@a cli-config}

{@a config}

## configuration

> A file named `angular.json` at the root level of an Angular [workspace](#workspace) provides workspace-wide and project-specific configuration defaults for build and development tools provided by the [Angular CLI](#cli).

Reviewer: ```suggestion
A file named `angular.json` at the root level of an Angular [workspace](#workspace) provides workspace-wide and project-specific configuration defaults for build and development tools that are provided by, or integrate with, the [Angular CLI](#cli).
```

Author: Yes--  I'm doing it in my name rather than committing the suggestion (to avoid CLA problems)

Reviewer: The CLA warning isn't an issue, any Googler who is doing to the final merge can dismiss it.

For more information, see [Workspace Configuration](guide/workspace-config).

> Additional project-specific configuration files are used by subsystems and tools, such as `package.json` for the [npm package manager](#npm-package), and `tsconfig.json` for [TypeScript transpilation](#transpile), and `tslint.json` for [TSLint](https://palantir.github.io/tslint/).

Reviewer: ```suggestion
Additional project-specific configuration files are used by tools, such as `package.json` for the [npm package manager](#npm-package), `tsconfig.json` for [TypeScript transpilation](#transpile), and `tslint.json` for configuring lint rules with [TSLint](https://palantir.github.io/tslint/).
```

## pipe

A class which is preceded by the `@Pipe{}` decorator and which defines a function that transforms input values to output values for display in a [view](guide/glossary#view). Angular defines various pipes, and you can define new pipes.

To learn more, see [Pipes](guide/pipes).

{@a polyfill}

## polyfill

An [npm package](guide/npm-packages) that plugs gaps in a browser's JavaScript implementation.
See [Browser Support](guide/browser-support) for polyfills that support particular functionality for particular platforms.

{@a project}

## project

> In Angular, a standalone application or [library](guide/glossary#library) in a [workspace](#workspace).

Reviewer 3: Hmmm. One thing I liked about the original/previous definition is that it makes it clear that project is a dev-time construct. An app or library exists in both dev-time and runtime contexts. The app or library is what you're developing; a project is the context in which you do that development. It's the set of source files and related config files needed to dev, test?, and deploy the app or library.

## output

> When defining a [directive](guide/glossary#directive), the `@Output{}` decorator on a directive property

Author: Done - Don't want to confuse the CLA

## project

> In the Angular development environment, a standalone application or [library](#library) in a [workspace](#workspace).

Reviewer: I think that this is too general. I suggested (not accepted) the addition of "Angular CLI" here because that is where `project` is relevant. It is possible to have Angular development environments that do not use the CLI and do not have the idea of `projects` or `workspaces`.

Author: How about this.

Reviewer: > In the Angular CLI, a standalone application or [library](#library) that can be created or modified by a CLI command.

Looks good.

> An application project (and its related end-to-end test project) contains the set of source files, resources, and configuration files that you need to develop, test, and deploy applications.

Reviewer: There is no `(and its related end-to-end test project)` in version 8.

> An application project (and its related end-to-end test project) contains the set of source files, resources, and configuration files that you need to develop, test, and deploy applications.

Reviewer: A application `project` does not include information for how to deploy an application. That is at the `workspace` level.

## project

> In the Angular development environment, a standalone application, test application, or [library](guide/glossary#library) in a [workspace](#workspace).

Reviewer: ```suggestion
In the Angular [CLI](#cli) development environment, a standalone application or [library](#library) in a [workspace](#workspace).
```

> A workspace can contain multiple projects, which share a compilation context.

Reviewer: ```suggestion
```

Author: Why remove this? Shared compilation space was the first definition I heard for "workspace"

Reviewer: This is the definition of `project` and the definition of `workspace` is linked above. This helps keep the definition of workspace in one place rather than needing to fix/change the definition here and there.

Author: Ah. Ok - I did finally get down to workspace and saw it there...

> A application project (and its related end-to-end test project) contains the set of source files, resources, and configuration files that you need to develop, test, and deploy applications.

Reviewer: ```suggestion
```

Author: Why remove this?

Reviewer: `(and its related end-to-end test project)` is no longer accurate for Angular CLI version 8 AFAIK. The rest of it is also not completely accurate either. It isn't needed to define what a CLI "project" is. Plus in the next line we point them to the Project File Structure guide which includes a more detailed explanation.

* The `complete` notification doesn't send a value, but the handler is called when the call completes. Scheduled values can continue to be returned after the call completes.

{@a T}

{@a target}

## target

Specifies the desired end-result of a complex process initiated by the [Architect](#architect) tool.

In the [workspace configuration file](guide/workspace-config#project-tool-configuration-options), an "architect" target object contains a configuration to use in creating that target for a particular project.

> Most Architect targets correspond to [CLI commands](#cli), such as `build`, `serve`, `test`, and `lint`.

Reviewer 3: I like how this focuses on the end result. How does that mesh with the description of ng run, where the target is the config file itself? Maybe just refine that doc as a follow-on step to match this idea of target as destination, and that command needs the config file associated with that destination/end result?

Author: I thought that was how I describe the run command - please keep track of this, we should check.

> Most Architect targets correspond to [CLI commands](#cli), such as `build`, `serve`, `test`, and `lint`.

Reviewer 3: Positioning: Link to Architect definition. Mention that ng run makes it possible to use custom builders and targets, as a parallel to the internal pre-defined targets mentioned here (build, serve, test, lint)?

{@a rule}

## rule

> In [schematics](#schematic), a function that operates on a [file tree](#file-tree) to create, delete, or modify code files in a specific manner.

Reviewer: ```suggestion
In [schematics](#schematic), a function that operates on a [file tree](#file-tree) to create, delete, or modify files in a specific manner.
```

A scaffolding library that defines how to generate or transform a programming project by creating, modifying, refactoring, or moving files and code.
A schematic defines [rules](#rule) that operate on a virtual file system called a [tree](#file-tree).

The [Angular CLI](guide/glossary#cli) uses schematics to generate and modify [Angular projects](guide/glossary#project) and parts of projects.

* Angular provides a set of schematics for use with the CLI. See the [Angular CLI command reference](cli). The [`ng add`](cli/add) command runs schematics as part of adding a library to your project. The [`ng generate`](cli/generate) command runs schematics to create apps, libraries, and Angular code constructs.

> * [Library](#library) developers can create schematics that enable the Angular CLI to add and update their published libraries, and to generate artifacts the library defines.

Reviewer: ```suggestion
* [Library](#library) developers can create schematics that [enable the Angular CLI](guide/creating-libraries#integrating-with-the-cli) to add and update their published libraries, and to generate artifacts the library defines.
```

npm install -g @angular-devkit/schematics-cli
</code-example>

> This installs the `schematics` executable, which you can use to create a new schematics [collection](#collection) with an initial named schematic. The collection folder is a workspace for schematics. You can use the CLI to add a new schematic to an existing collection, or extend an existing schematic.

Reviewer: ```suggestion
This installs the `schematics` executable, which you can use to create a new schematics [collection](#collection) with an initial named schematic. The collection folder is a workspace for schematics. You can use the `schematics` executable to add a new schematic to an existing collection, or extend an existing schematic.
```

## structural directives

A category of [directive](#directive) that is responsible for shaping HTML layout by modifying the DOM&mdashthat is, adding, removing, or manipulating elements and their children.

To learn more, see [Structural Directives](guide/structural-directives).

{@a subscriber}

## subscriber

A function that defines how to obtain or generate values or messages to be published. This function is executed when a consumer calls the `subscribe()` method of an [observable](#observable).

The act of subscribing to an observable triggers its execution, associates callbacks with it, and creates a `Subscription` object that lets you unsubscribe.

The `subscribe()` method takes a JavaScript object (called an [observer](#observer)) with up to three callbacks, one for each type of notification that an observable can deliver:

* The `next` notification sends a value such as a number, a string, or an object.
* The `error` notification sends a JavaScript Error or exception.
* The `complete` notification doesn't send a value, but the handler is called when the call completes. Scheduled values can continue to be returned after the call completes.

{@a T}

{@a target}

## target

> An object in the [workspace configuration file](guide/workspace-config#project-tool-configuration-options) that configures an [Architect](#architect) [builder](#builder) for use in a [project](#project) context.

Reviewer 2: ```suggestion
A buildable or runnable subset of a [project](#project), configured as an object in the [workspace configuration file](guide/workspace-config#project-tool-configuration-options), and executed by an [Architect](#architect) [builder](#builder).
```

* The `complete` notification doesn't send a value, but the handler is called when the call completes. Scheduled values can continue to be returned after the call completes.

{@a T}

{@a target}

## target

An object in the [workspace configuration file](guide/workspace-config#project-tool-configuration-options) that configures an [Architect](#architect) [builder](#builder) for use in a [project](#project) context.

> In the `angular.json` file, each project has an "architect" section that contains targets which configure built-in builders. Some of these targets correspond to [CLI commands](#cli), such as `build`, `serve`, `test`, and `lint`.

Reviewer: I think that `built-in` here should be removed as these are not limited to builders provided by Architect and this does not specify that it is only talking about the state of the configuration in a newly generated project.

> For example, the Architect builder invoked by the `ng build` command to compile a project uses a particular build tool, and has a default configuration whose values can be overridden on the command line. The `build` target also defines an alternate configuration for a "production" build, that can be invoked with flag on the `build` command.

Reviewer: There seems to be a grammar issue here:

>  that can be invoked with flag on the `build` command.

> The [CLI](#cli) provides a set of builders. The [`ng new` command](cli/new) provides a set of targets for the initial application project. The [`ng generate application`](cli/generate#application) and [`ng generate library`](cli/generate#library) commands provide a set of targets for each new [project](#project). These targets, their options and configurations, can be customized to meet the needs of your project. For example, you may want to add a "staging" or "testing" configuration to a project's "build" target.

Reviewer: ```suggestion
The [Architect](#architect) provides a set of builders. The [`ng new` command](cli/new) provides a set of targets for the initial application project. The [`ng generate application`](cli/generate#application) and [`ng generate library`](cli/generate#library) commands provide a set of targets for each new [project](#project). These targets, their options and configurations, can be customized to meet the needs of your project. For example, you may want to add a "staging" or "testing" configuration to a project's "build" target.
```

* The `complete` notification doesn't send a value, but the handler is called when the call completes. Scheduled values can continue to be returned after the call completes.

{@a T}

{@a target}

## target

A specific default configuration for an [Architect](#architect) [builder](#builder).

In the [workspace configuration file](guide/workspace-config#project-tool-configuration-options), a   builder object in the "architect" section contains targets which configure built-in builders that correspond to [CLI commands](#cli), such as `build`, `serve`, `test`, and `lint`.

For example, the Architect builder invoked by the `ng build` command to compile a project uses a particular build tool, and has a default configuration whose values can be overridden on the command line. There is also an optional "production" target specified in the "build" subsection of "architect" in the `angular.json` file. The target object provides overrides for some options; those values can in turn be overridden on the command line.

> There are some additional predefined builders and targets, and it is possible to define your own custom builders and target configurations. You can run a builder with a specified target using the [`ng run`](cli/run) CLI command.

Reviewer: ```suggestion
The [CLI](#cli) provides a set of builders. The [`ng new` command](cli/new) provides a set of targets for the initial application project. The [`ng generate application`](cli/generate#application) and [`ng generate library`](cli/generate#library) commands provide a set of targets for each new [project](#project). These targets, their options and configurations, can be customized to meet the needs of your project. For example, you may want to add a "staging" or "testing" configuration to a project's "build" target.

Additionally, it is possible to define a custom builder. You can define a new target, in a project, that uses your custom builder. Then you can run that target using the [`ng run`](cli/run) CLI command.
```

## web component

See [custom element](#custom-element).

{@a workspace}

## workspace

> An compilation context used by the [Angular CLI](#cli) for [projects](#project) (that is, applications and libraries).

Reviewer 2: ```suggestion
A collection of Angular [projects](#project) (that is, applications and libraries) powered by the [Angular CLI](#cli) that are typically collocated in a single source control repository (e.g. [git](https://git-scm.com/)).
```

## workspace

> An Angular compilation context for [projects](guide/glossary#project) (that is, applications and libraries).

Reviewer: ```suggestion
An Angular compilation context for [projects](#project) (that is, applications and libraries).
```

Author: If we do this one, we might as well do all of them.

> An Angular compilation context for [projects](guide/glossary#project) (that is, applications and libraries).

Reviewer: `workspace`s are specific to the Angular CLI. It is possible to develop Angular applications without the Angular CLI. So I think that the following is too broad

> An Angular compilation context
